base/abstractarraymath.jl

```diff
@@ -330,89 +308,171 @@ julia> repeat([1 2; 3 4], inner=(2, 1), outer=(1, 3))
+end
+
+function check(arr, inner, outer)
+    if inner != nothing
+        if any(<(0), inner)
+            msg = """
+            Error calling repeat. all(inner . >= 0) must hold. Got
+            inner = $inner
+            """
+            throw(ArgumentError(msg))
+        end
+        if length(inner) < ndims(arr)
+            msg = """
+            Error calling repeat. length(inner) >= ndims(arr) must hold. Got
+            inner = $inner
+            size(arr) = $(size(arr))
+            """
+            throw(ArgumentError(msg))
+        end
     end
+    if outer != nothing
+        if any(<(0), outer)
+            msg = """
+            Error calling repeat. all(outer . >= 0) must hold. Got
+            outer = $outer
+            """
+            throw(ArgumentError(msg))
+        end
+        if (length(outer) < ndims(arr)) && (inner != nothing)
+            msg = """
+            Error calling repeat. length(outer) >= ndims(arr) must hold. Got
+            outer = $outer
+            size(arr) = $(size(arr))
+            """
+            throw(ArgumentError(msg))
+        end
+    end
+end
 
-    # fill the first inner block
-    if all(isequal(1), inner)
-        idxs = (axes(A)..., ntuple(n->OneTo(1), ndims(R)-ndims(A))...) # keep dimension consistent
-        R[idxs...] = A
-    else
-        inner_indices = [1:n for n in inner]
-        for c in CartesianIndices(axes(A))
-            for i in 1:ndims(A)
-                n = inner[i]
-                inner_indices[i] = (1:n) .+ ((c[i] - 1) * n)
-            end
-            fill!(view(R, inner_indices...), A[c])
+repeat_inner_outer(arr, inner::Nothing, outer::Nothing) = arr
+repeat_inner_outer(arr, ::Nothing, outer) = repeat_outer(arr, outer)
+repeat_inner_outer(arr, inner, ::Nothing) = repeat_inner(arr, inner)
+repeat_inner_outer(arr, inner, outer) = repeat_outer(repeat_inner(arr, inner), outer)
+
+function repeat_outer(a::AbstractMatrix, (m,n)::NTuple{2, Any})
+    o, p = size(a,1), size(a,2)
+    b = similar(a, o*m, p*n)
+    for j=1:n
+        d = (j-1)*p+1
+        R = d:d+p-1
+        for i=1:m
+            c = (i-1)*o+1
+            @inbounds b[c:c+o-1, R] = a
         end
     end
+    return b
+end
 
-    # fill the outer blocks along each dimension
-    if all(isequal(1), outer)
-        return R
+function repeat_outer(a::AbstractVector, (m,)::Tuple{Any})
+    o = length(a)
+    b = similar(a, o*m)
+    for i=1:m
+        c = (i-1)*o+1
+        @inbounds b[c:c+o-1] = a
     end
-    src_indices  = [1:n for n in inner_shape]
-    dest_indices = copy(src_indices)
-    for i in eachindex(outer)
-        B = view(R, src_indices...)
-        for j in 2:outer[i]
-            dest_indices[i] = dest_indices[i] .+ inner_shape[i]
-            R[dest_indices...] = B
+    return b
+end
+
+function repeat_outer(arr::AbstractArray{<:Any,N}, dims::NTuple{N,Any}) where {N}
+    repeat_outer_generic(arr, dims)
+end
+
+function repeat_outer_generic(arr, dims)
```

Comment:
Is there a reason for the separate function here?

```suggestion
```

Reply:
It was useful for checking that the 1d and 2d special implementations are needed for performance, but is no longer necessary.